### PR TITLE
add offline-render utils

### PR DESCRIPTION
**Pull request by eundersander**
## Motivation and Context

The goal of these utils is to help a user produce renders (images) given a user-specified model, camera intrinsics and camera transforms.

Steps shown in the demo function:
1. Write a Habitat replay file that stores (a) a reference to our desired model to be rendered, and (b) a camera transform (position and rotation) for each render.
2. Initialize the simulator for offline rendering, specifying camera intrinsics.
3. Load the replay and produce renders (saved as png files).

The use of a Habitat replay file is slightly overkill, but it'll help us extend to more complex scenes (multiple models) and other renderers (Unity, Blender) in the future.

## Texture Downsampling
Update March 2: support for texture downsampling, see images in comments below. Think of texture resolution as "texels per square meter" in world space. downsample_factor=0 is default and full resolution. 1 means downsampling 2x, e.g. a MP3D 2048x2048 texture gets downsampled to 1024x1024. As the downsample_factor increases and the downsampled texture resolutions approach 1x1, the the visuals start to look like solid-colored shapes with no texture.

## Mesh Simplification
Update April 16: support for mesh simplification at load-time, see images below. This uses the Magnum MeshOptimizer plugin. We reduce mesh triangle count to some fraction of the original count. As mesh_simplification_fraction drops below 0.5, we see more and more severe artifacts like gaps in the mesh.

## How Has This Been Tested

todo: add python unit tests for these utilities

## Types of changes

- [ ] Docs change / refactoring / dependency upgrade
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist

- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [x] I have completed my CLA (see **CONTRIBUTING**)
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

**Comment by eundersander** [comment]
Texture downsampling on MP3D scenes:
Z6MFQCViBuw_skip0 0
Z6MFQCViBuw_skip2 0
Z6MFQCViBuw_skip4 0

2t7WUuJeko7_skip0
2t7WUuJeko7_skip2
2t7WUuJeko7_skip4

**Comment by eundersander** [comment]
## Mesh Simplification
The green wireframe is visualized here to help understanding (in the C++ viewer, use `--enable-physics --debug-bullet` flags).

mp3d_simplify_1
mp3d_simplify_2
mp3d_simplify_3
mp3d_simplify_2t7WUuJeko7_1
mp3d_simplify_2t7WUuJeko7_2
mp3d_simplify_2t7WUuJeko7_3

File: src/utils/viewer/viewer.cpp [hunk 1]
```diff
@@ -524,6 +525,10 @@ Viewer::Viewer(const Arguments& arguments)
     Mn::Debug{} << "Stage using DEFAULT_LIGHTING_KEY";
     simConfig.sceneLightSetup = esp::DEFAULT_LIGHTING_KEY;
   }
+  if (!args.value("mesh-simplification").empty()) {
+    simConfig.meshSimplificationFraction =
+        (float)std::stoi(args.value("mesh-simplification")) / 100;
```
Comment (mosra) [inline src/utils/viewer/viewer.cpp:530]: FYI, you can do this:

```suggestion
        args.value<Float>("mesh-simplification") / 100;
```

File: src/esp/assets/GenericMeshData.cpp [hunk 2]
```diff
@@ -73,11 +81,49 @@ void GenericMeshData::setMeshData(Magnum::Trade::MeshData&& meshData) {
...
+            manager.loadAndInstantiate("MeshOptimizerSceneConverter"));
+
+    meshoptimizer->configuration().setValue("simplify", "true");
+    meshoptimizer->configuration().setValue(
+        "simplifyTargetIndexCountThreshold",
+        std::to_string(targetMeshSimplificationFraction));
+    constexpr float simplifyTargetError =
+        0.2;  // I don't think this field has any effect
+    meshoptimizer->configuration().setValue(
+        "simplifyTargetError", std::to_string(simplifyTargetError));
```
Comment (mosra) [inline src/esp/assets/GenericMeshData.cpp:106]: ...and same here (Configuration and Arguments use the same internals for type conversion):

```suggestion
        "simplifyTargetError", simplifyTargetError);
```

**Review by mosra — COMMENTED** [review]
> As mesh_simplification_fraction drops below 0.5, we see more and more severe artifacts like gaps in the mesh.

IIRC this got substantially improved in [meshoptimizer 0.16](https://github.com/zeux/meshoptimizer/releases/tag/v0.16), when I get the time I'll adapt the plugin to work with it and ping you here. Ideally we should get to about a 10x reduction without major artifacts.

**Comment by mosra** [comment]
Support for meshoptimizer 0.16 is available since #1329.

**Pull request closed without merging**